### PR TITLE
Transplant the change of scikit-learn into scikit-image for RANSAC 

### LINKED ISSUE #5952: RANSAC: Check for updates
### Discussed in https://github.com/scikit-image/scikit-image/discussions/5938

<div type='discussions-op-text'>

<sup>Originally posted by **stefanv** July 11, 2017</sup>
RANSAC was implemented in scikit-image & scikit-learn around the same time, by the same @ahojnnes.  In the mean time, some improvements have landed on either side.

We should compare the two implementations, and port fixes to either side as needed.
</div>

### PR DESCRIPTION
**Pull request by gmelon**
## Description

Fixes #5952 

I compared the code implementation of the RANSAC model of scikit-image & scikit-learn.

What's in common is that both models have hyperparameter “max_trials”, and use the _dynamic_max_trials function to commute dynamic max trials for RANSAC algorithm.

Originally, scikit-image's RANSAC compare whether _dynamic_max_trials are less than max_trials only when the number of inliers_count is greater than best_inlier_num. And most of the _dynamic_max_trials calculated each time have infinity. The result below is the output of the corresponding value whenever _dynamic_max_trials are calculated when RANSAC is performed with the original code. And lastly, I printed out the total number of Iterations.
```
inf
inf
inf
inf
inf
inf
inf
inf
Iteration : 1000
```

So, when calculating _dynamic_max_trials, we set the value to max_trials if it is less than the existing max_trials.  And we modified the code by referring to scikit-learn so that the actual value is calculated more frequently than infinity when calculating the _dynamic_max_trials. Below is the revised result.
```
5983.0
166.0
138.0
133.0
Iteration : 133
```
You can see that _dynamic_max_trials actually decrease and the total number of iterations decreases dramatically. Using the sample code of scikit-learn, we confirmed that the results generated in the two cases were also the same when visually checked. The products in both cases are as follows.
original
modified

In summary, I have equalized both the code implementation that judges max_trials and the _dynamic_max_trials function to scikit-learn code style.

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/measure/fit.py
Comment (jni) [inline skimage/measure/fit.py:615]:
```suggestion
    inlier_ratio = n_inliers / n_samples
```

This was a Python 2-ism. :joy:

File: skimage/measure/fit.py
Comment (jni) [inline skimage/measure/fit.py:621]:
```suggestion
        return np.inf
```

File: skimage/measure/fit.py
Comment (jni) [inline skimage/measure/fit.py:841]:
```suggestion
    num_trials = 0
    # max_trials can be updated inside the loop, so this cannot be a for-loop
    while num_trials < max_trials:
        num_trials += 1
```

I found this confusing when I first read it and was about to ask you to revert to a for-loop, so I think leaving a comment here for future readers will be useful.

**Review by jni — APPROVED** [review]
Thanks @gmelon! I've left some minor nitpicky comments, but otherwise this looks great to me! The Azure failures appear to be related to an issue with Azure itself rather than this PR.

**Comment by gmelon** [comment]
@jni , Thank you for your kind review.
I was worried that the build problem was mine, but I'm glad.
As you said, I revised the code and commited it, so please check it.

File: skimage/measure/fit.py
Comment (rfezzani) [inline skimage/measure/fit.py:621]:
```suggestion
    if probability == 0:
        return 0
    if n_inliers == 0:
        return np.inf
    inlier_ratio = n_inliers / n_samples
    nom = max(_EPSILON, 1 - probability)
    denom = max(_EPSILON, 1 - inlier_ratio ** min_samples)
```
`_dynamic_max_trials` being called in the loop, this may save some FLOPS :slightly_smiling_face:

**Review by rfezzani — COMMENTED** [review]
Thank you @gmelon, I just left a small suggestion, otherwise it is all good :wink:

**Comment by gmelon** [comment]
Thank you @rfezzani , I just modified that part and commited it, so I'd appreciate it if you could check it. 😆

File: skimage/measure/fit.py
Comment (rfezzani) [inline skimage/measure/fit.py:620]: I don't know if it worth it, but shouldn't we also keep this special cases?

Comment (gmelon) [reply skimage/measure/fit.py:620]: I think it's necessary.
By the way, if the default value of probability is 1, the effect of dynamic maximum decreases, so what do you think about adding that code and adjusting the default value of probability from 1 to 0.99.

File: skimage/measure/fit.py
Comment (rfezzani) [inline skimage/measure/fit.py:622]: I think that the use of `abs` is not necessary: both `nom` and `denom` are in `[_EPSILON, 1]`, so both `log` values are negative and theire ratio is positive. Right?

File: skimage/measure/fit.py
Comment (rfezzani) [reply skimage/measure/fit.py:620]: Changing the default value is a breaking change and needs starting a deprecation cycle. We need more @scikit-image/core inputs to take this decision.

File: skimage/measure/fit.py
Comment (gmelon) [reply skimage/measure/fit.py:622]: I agree on it. I'll remove `abs`.

File: skimage/measure/fit.py
Comment (gmelon) [reply skimage/measure/fit.py:620]: Can you explain what depression cycle is? I'm Sorry :(
Additionally, do you think it's better to proceed with that process? Or do you think it's better to just add the code above and maintain the default value?

Comment (rfezzani) [reply skimage/measure/fit.py:620]: You can find in the [contribution guide](https://scikit-image.org/docs/dev/contribute.html#deprecation-cycle) the strategy that `skimage` adopts to manage any behavior change in the library.
I have no strong opinion about what the default value should be, but I see that `sklearn` uses 0.99 as default.

Comment (gmelon) [reply skimage/measure/fit.py:620]: Thank you. I checked contribution guide and sklearn's PR when _dynamic_max_trials were first implemented.
If the results when _dynamic_max_trials are applied are not different from the existing results and the speed increases, I think it is okay to set the stop_probability to the default value of 0.99 which is large enough but not 1.

`We need more @scikit-image/core inputs to take this decision.`
-> Does this mean we need to get opinions from other maintainers?

Comment (rfezzani) [reply skimage/measure/fit.py:620]: > -> Does this mean we need to get opinions from other maintainers?

Yes, that's what I meant :wink:

Comment (gmelon) [reply skimage/measure/fit.py:620]: Thank you. Is there anything I need to do to get an opinion? Is it enough to just wait?
Because I can't stand my commit being merged. It's so excited😆

Comment (rfezzani) [reply skimage/measure/fit.py:620]: We simply need to wait now: core devs are already notified, just give them some time to answer :wink:

Comment (gmelon) [reply skimage/measure/fit.py:620]: Okay, Thanks a lot. 🙂

Comment (stefanv) [reply skimage/measure/fit.py:620]: I think this is fine to put in without deprecation.

File: skimage/measure/fit.py
Comment (stefanv) [inline skimage/measure/fit.py:838]: Does num_trials get updated inside the loop? Doesn't look like it, so this was probably fine?

Comment (gmelon) [reply skimage/measure/fit.py:838]: @stefanv yes, num_trials is updated at line 840.

```
839 while num_trials < max_trials:
840    num_trials += 1
```

File: skimage/measure/fit.py
Comment (gmelon) [reply skimage/measure/fit.py:620]: @stefanv
So you're saying we should add the code above to _dynamic_max_trials and keep the default stop_probability value of ransac() at 1, right?

File: skimage/measure/fit.py
Comment (stefanv) [reply skimage/measure/fit.py:838]: Yes, but if we use the old code that line is not there or necessary.

File: skimage/measure/fit.py
Comment (stefanv) [reply skimage/measure/fit.py:620]: Why does sklearn use 0.99?

File: skimage/measure/fit.py
Comment (gmelon) [reply skimage/measure/fit.py:838]: In my opinion, since max_trials can be changed in the new code(because of the result of _dynamic_max_trials can update max_trials), we need to record and increase num_trials separately, so we need that row.
Am I misunderstanding your intentions?😢

Comment (stefanv) [reply skimage/measure/fit.py:838]: Ah, no, that makes sense! I was looking at num_trials

File: skimage/measure/fit.py
Comment (gmelon) [reply skimage/measure/fit.py:620]: In my opinion, if it is set to 1, the dynamic maximum is not calculated, so it seems that the value is set to 0.99, so that the dynamic maximum is calculated without affecting the result as much as possible.
In other words, I think this is just the difference in decision whether to use dynamic maximum or not.
What do you think about it @stefanv ?

Comment (gmelon) [reply skimage/measure/fit.py:620]: @stefanv @rfezzani 
The default value for the probability of skimage is set to 1. Therefore, the existing skimage code always returns the infinite number of trials. I think this is an inefficient way.
Since the commonly used value is 0.99, it is reasonable to set the default to 0.99 **if the user does not want to pay particular attention** to probability. In the following paper, 0.99 is used as the value of RANSAC's confidence probability. [FestGPU: a framework for fast robust estimation on GPU](https://www.researchgate.net/figure/The-influence-of-the-RANSAC-parameters-to-the-number-of-iterations-p-099-For-larger_fig6_268124277)
And like @stefanv said, I think the fact that the default value of probability in sklearn is 0.99 can be the basis for changing the default value of skimage.

Comment (gmelon) [reply skimage/measure/fit.py:620]: And from a different point of view, even if the default value is not modified, I think it will have a similar effect to the above discussion if the code compares the value that goes into the log as below with epsilon.
```
nom = max(_EPSILON, 1 - probability)
denom = max(_EPSILON, 1 - inlier_ratio ** min_samples)
return float(np.ceil(np.log(nom) / np.log(denom)))
```

In the same context, we don't need the code below, do we?
```
if probability == 1:
    return np.inf

if n_inliers == n_samples:
    return 1
```

Comment (rfezzani) [reply skimage/measure/fit.py:620]: I think that we can remove the handling of these special cases.

File: skimage/measure/fit.py
Comment (rfezzani) [reply skimage/measure/fit.py:622]:
```suggestion
    return np.ceil(np.log(nom) / np.log(denom))
```

**Review by rfezzani — COMMENTED** [review]
Sorry @gmelon for very late answer :pray:

**Comment by gmelon** [comment]
> Sorry @gmelon for my very late answer 🙏

It's okay😉 Thank you for the decision.
I committed the changes, so please check it out!

File: skimage/measure/fit.py
Comment (rfezzani) [reply skimage/measure/fit.py:620]: The test for `probability == 1` is necessary for the tests to pass :)

Comment (gmelon) [reply skimage/measure/fit.py:620]: @rfezzani Oh, I see. I modified it!

File: skimage/measure/tests/test_fit.py
Comment (rfezzani) [inline skimage/measure/tests/test_fit.py:375]: `_dynamic_max_trial` must return a positive number :confused:

Comment (gmelon) [reply skimage/measure/tests/test_fit.py]: Oh no! I modified it.
First, I adjusted the number of samples to 10 at the beginning, but I returned it to 5 again.
And since we deleted the condition of unconditionally outputting inf when probability was 1, a very large value would be output, so I put that value as a condition.
What about this?

Comment (rfezzani) [reply skimage/measure/tests/test_fit.py]: :+1: Let's wait for the CI to be green again ;)

Comment (gmelon) [reply skimage/measure/tests/test_fit.py]: > 👍 Let's wait for the CI to be green again ;)

The results came out. What do you think?
From what I see, there seems to be no problem with this PR.

Comment (gmelon) [reply skimage/measure/tests/test_fit.py]: > 👍 Let's wait for the CI to be green again ;)

We were done! 😁

**Review by rfezzani — APPROVED** [review]
Thank you for your patience @gmelon, merging :wink:

**Pull request merged**
